### PR TITLE
update setupt.f90, update_guess.f90 etc based on latest GSL 2mDA code

### PR DESCRIPTION
**Pull request by guoqing-noaa**

### CODE REVIEW

File: src/gsi/gsd_update_mod.f90 [hunk 1]
```diff
@@ -552,17 +553,15 @@ subroutine gsd_update_th2(tinc,it)
               if(ihaveq/=0) cycle
               dth2=tinc(i,j)/(one+fv*ges_q(i,j,1))
```
Comment (shlyaeva) [inline src/gsi/gsd_update_mod.f90:554]: Not familiar with this code, but looks like in this case it maybe uses the lowest level of 3D background humidity: is that intentional?

File: src/gsi/setupt.f90 [hunk 2]
```diff
@@ -752,7 +741,16 @@ subroutine setupt(obsLL,odiagLL,lunin,mype,bwork,awork,nele,nobs,is,conv_diagsav
...
+          if(save_jacobian) then
+             t2m_ind = getindex(svars2d, 't2m')
+             if (t2m_ind < 0) then
+                 print *, 'Error: no variable t2m in state vector.Exiting.'
+                 call stop2(1300)
+             endif
+             dhx_dx%st_ind(1) = sum(levels(1:ns3d))  + t2m_ind
+             dhx_dx%end_ind(1) = sum(levels(1:ns3d)) + t2m_ind
+             dhx_dx%val(1) = one
+          endif
```
Comment (shlyaeva) [inline src/gsi/setupt.f90:753]: Looks correct to me; I agree that this should be almost exactly the same as the 2D `ps` case where observation operator only does horizontal interpolation (except for the variable name).

**Review by shlyaeva — COMMENTED** [review]
A cursory review since I'm not very familiar with the 2m update. I carefully reviewed the part for saving Jacobian, and only looked through the rest, looks good to me! Didn't test.

File: src/gsi/gsd_update_mod.f90 (hunk 1 above)
Comment (guoqing-noaa) [reply src/gsi/gsd_update_mod.f90:554]: @shlyaeva You are right. If the increment is in the virtual temperature form, it still uses the lowest level of 3D background humidity.  I think currently we assume that the increments are in the sensible form. I will double check this. Thanks!

File: src/gsi/setupt.f90 [hunk 3]
```diff
@@ -805,13 +794,12 @@ subroutine setupt(obsLL,odiagLL,lunin,mype,bwork,awork,nele,nobs,is,conv_diagsav
 
 ! SCENARIO 4: obs is sfctype, and i_use_2mt4b flag is on (turns on LAM sfc DA)
         if(i_use_2mt4b>0 .and. sfctype) then
+
            if(i_coastline==1 .or. i_coastline==3) then
 
 !          Interpolate guess th 2m to observation location and time
-              call tintrp2a11_csln(ges_th2,tges2m,tges2m_water,dlat,dlon,dtime,hrdifsig,&
```
[the remark this answers is not in the document]
Comment (guoqing-noaa) [reply src/gsi/setupt.f90:811]: No, right now i_use_2mt4b is for regional and do_global_2mDA is for global. 
Don't worry about the regional part. We already made all changes from th2 to t2m for the regional part and the changes were already in the GSI master branch.

File: src/gsi/update_guess.f90 [hunk 4]
```diff
@@ -454,15 +454,15 @@ subroutine update_guess(sval,sbias)
         endif
         call  gsd_update_soil_tq(tinc_1st,is_t,qinc_1st,is_q,it)
      endif  ! l_gsd_soilTQ_nudge
-     if (i_use_2mt4b > 0 .and. is_t>0) then
+     if ( (i_use_2mt4b > 0.or.do_global_2mDA) .and. is_t>0) then
```
[the remark this answers is not in the document]
Comment (guoqing-noaa) [reply src/gsi/update_guess.f90:457]: When I added this, I was not aware of your separate treatment.  We will look what you added and see how it goes forward.

Regarding to this part, it works for both regional and global as gsd_update_t2m is simply to update ges_t2m  (ges_t2m(i,j) = ges_t2m(i,j) + dth2) which is same for the regional and global.

**Pull request merged**